### PR TITLE
Rewards computation for airdrop

### PR DESCRIPTION
**Pull request by raamb**
- Create snapshot from balances DB everyday during snapshot period
- Created rewards processor for Nunet
- Compute rewards for each user (needs to be comlpeted)

### CODE REVIEW

File: airdrop/alembic/versions/5353f0a8d078_.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+
+# revision identifiers, used by Alembic.
+revision = '5353f0a8d078'
+down_revision = ('1c6bb2fd25d5', '7fdcc2e22f9b')
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
```
Comment (prashantramangupta) [inline airdrop/alembic/versions/5353f0a8d078_.py:19]: This file does not serve any purpose.

File: airdrop/infrastructure/models.py [hunk 2]
```diff
@@ -145,9 +150,32 @@ class UserReward(Base, AuditClass):
     )
     address = Column("address", VARCHAR(50), nullable=False, index=True)
     condition = Column("condition", TEXT, nullable=True)
-    rewards_awarded = Column("rewards_awarded", INTEGER, nullable=False)
+    rewards_awarded = Column("rewards_awarded", BIGINT, nullable=False)
+    score = Column("score", DECIMAL(10,5), nullable=False)
+    log10_score = Column("log10_score", DECIMAL(10,8), nullable=False)
```
Comment (prashantramangupta) [inline airdrop/infrastructure/models.py:155]: We should not have a log10_score column? We should keep a generic name like the eligible amount. This name is specific to the airdrop portal where this formula is applicable.

File: airdrop/infrastructure/models.py [hunk 3]
```diff
@@ -115,6 +117,9 @@ class UserBalanceSnapshot(Base, AuditClass):
     )
     address = Column("address", VARCHAR(50), nullable=False)
     balance = Column("balance", BIGINT, nullable=False)
+    staked = Column("staked", BIGINT, nullable=False)
+    total = Column("total", BIGINT, nullable=False)
+    snapshot_guid = Column("snapshot_guid", VARCHAR(50), nullable=False)
```
Comment (prashantramangupta) [inline airdrop/infrastructure/models.py:122]: If guid type in hex then we have to increase the limit to 68

File: airdrop/infrastructure/models.py [hunk 4]
```diff
@@ -105,7 +108,6 @@ class AirdropWindowTimelines(Base, AuditClass):
     date = Column("date", TIMESTAMP(), nullable=False)
```
Comment (prashantramangupta) [inline airdrop/infrastructure/models.py:108]: This column name is very generic

File: airdrop/job/eligibility.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+from airdrop.config import BALANCE_DB_CONFIG, NETWORK, SLACK_HOOK
+from common.logger import get_logger
+from pydoc import locate
+from decimal import Decimal
+
+logger = get_logger(__name__)
+
+class EligibilityProcessor:
+    def __init__(self):
+        self._airdrop_db = Repository(NETWORK["db"])
```
Comment (prashantramangupta) [inline airdrop/job/eligibility.py:19]: Are we sure that we want to keep the DB connection at the object level?

File: airdrop/job/eligibility.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+            details["rewards_processor"] = window["rewards_processor"]
+            self._active_airdrop_window_map[window["airdrop_window_id"]] = details
+        
+    def __batch_insert(self, values, force=False):
+        start = time.process_time()
+        number_of_rows = len(self.__rows_to_insert)
+        if (force and number_of_rows > 0) or number_of_rows >= 50:
+            self._airdrop_db.bulk_query(self.__insert_snapshot, self.__rows_to_insert)
+            self.__rows_to_insert.clear()       
+            print(f"*****{(time.process_time() - start)} seconds. Inserted {number_of_rows} rows")
```
Comment (prashantramangupta) [inline airdrop/job/eligibility.py:45]: We should make use of logger instead of print.

File: airdrop/job/eligibility.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    e.process_eligibility()
+    
+    logger.info(f"Completed Processing eligibility")
+    return generate_lambda_response(
+        200,
+        "Success"
+    )
+
+
+process_eligibility(None, None)
```
Comment (prashantramangupta) [inline airdrop/job/eligibility.py:95]: Line 95 is for local testing. If we are going to run it as lambda then we don't need this line.

File: airdrop/infrastructure/models.py (hunk 3 above)
Comment (raamb) [reply airdrop/infrastructure/models.py:122]: Its UUID so 50 is plenty

File: airdrop/serverless.yml [hunk 8]
```diff
@@ -102,6 +102,26 @@ package:
     - "!node_modules/**"
 
 functions:
+  process_eligibility:
```
Comment (prashantramangupta) [inline airdrop/serverless.yml:105]: Lambda timeout is set to 30 sec. Are we good with this timeout config?

File: airdrop/infrastructure/models.py (hunk 2 above)
Comment (raamb) [reply airdrop/infrastructure/models.py:155]: Agree. Will update

File: airdrop/infrastructure/models.py (hunk 4 above)
Comment (raamb) [reply airdrop/infrastructure/models.py:108]: Already existing column, will let Rajeev look into this

File: airdrop/job/eligibility.py (hunk 5 above)
Comment (raamb) [reply airdrop/job/eligibility.py:19]: Yes, this is a batch job that runs on a schedule. At a later point we will remove the DB connection for Balances and switch this to Alembic

File: airdrop/job/eligibility.py (hunk 6 above)
Comment (raamb) [reply airdrop/job/eligibility.py:45]: Yeah, this is incomplete. Will clean this up when I complete the logic

File: airdrop/job/eligibility.py (hunk 7 above)
Comment (raamb) [reply airdrop/job/eligibility.py:95]: Yeah, this is incomplete. Will clean this up when I complete the logic

File: airdrop/serverless.yml (hunk 8 above)
Comment (raamb) [reply airdrop/serverless.yml:105]: No, will update this.

File: airdrop/infrastructure/models.py (hunk 4 above)
Comment (rajeevbbqq) [reply airdrop/infrastructure/models.py:108]: The date column is on the AidropWindowTimeline table, which represents the timeline date. Do we need to rename it?

**Review by prashantramangupta — APPROVED** [review]
Approved

**Pull request merged**